feat: add skip_macro_invocations option

tommilligan opened the pull request:
Tracking issue: https://github.com/rust-lang/rustfmt/issues/5346
Original feature request and design: https://github.com/rust-lang/rustfmt/issues/5324#issuecomment-1111797851

Globally skippable macro names are defined in the config, and added to the existing `SkipContext`. Thanks to @ytmimi for the pointers on where to implement this, it made the changes super painless!

ytmimi commented on Configurations.md:
````diff
@@ -1006,6 +1006,39 @@ macro_rules! foo {
+- **Possible values**: a list of macro name idents, `["name_0", "name_1", ...]`
+- **Stable**: No (tracking issue: [#5346](https://github.com/rust-lang/rustfmt/issues/5346))
+
+#### `[]` (default):
+
+```rust
+macro_rules! foo {
+    ($a: ident : $b: ty) => {
+        $a(42): $b;
+    };
+    ($a: ident $b: ident $c: ident) => {
+        $a = $b + $c;
+    };
+}
+```
````
Can we add a call to `foo!` in this example to show that it'll be formatted when the `skip_macro_list` is empty?

ytmimi commented on src/config/macro_names.rs:
```diff
@@ -0,0 +1,81 @@
+use serde_json as json;
+use thiserror::Error;
+
+/// Defines the name of an input - either a file or stdin.
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Deserialize, Serialize)]
+pub struct MacroName(String);
```
Is this doc comment correct? I think this is the doc comment for `FileName`. Can we update or remove this.

ytmimi commented on src/config/macro_names.rs:
```diff
@@ -0,0 +1,81 @@
+    Json(json::Error),
+}
+
+// This impl is needed for `Config::override_value` to work for use in tests.
+impl str::FromStr for MacroNames {
+    type Err = MacroNamesError;
+
+    fn from_str(s: &str) -> Result<MacroNames, Self::Err> {
+        json::from_str(s).map_err(MacroNamesError::Json)
+    }
+}
```
I believe the current `IgnoreList` for ignored files always raises an error in it's `FromStr` impl to disallow setting an ignore list via the command line using `--config`.

I don't feel strongly either way, but I wonder if we should also require that the `skip_macro_names` list only be provided via the `rustfmt.toml`?

tommilligan replied:
Pragmatically, I use this implementation in the `tests` to pass in the option for test cases - there is prior art for this with `--file-lines`.

I don't see particularly why you'd want to forbid setting this from the command line?

ytmimi commented on src/config/mod.rs:
```diff
@@ -69,6 +72,8 @@ create_config! {
     format_macro_matchers: bool, false, false,
         "Format the metavariable matching patterns in macros";
     format_macro_bodies: bool, true, false, "Format the bodies of macros";
+    skip_macro_names: MacroNames, MacroNames::default(), false,
+        "Skip formatting the bodies of macros invoked with the following names.";
```
Does this option prevent formatting of macro definition bodies and macro calls or just macro calls? Based on the original issue (#5324) and the test examples I would expect that this just prevents macro calls from being formatted. Maybe we should update the description to make that more clear.

tommilligan replied:
Happy to update to be clearer - I also think the name of the option itself could be clearer, after realising the potential sources of confusion.

Potential options:
- skip_macro_invocations
- skip_macro_calls

I'd personally vote for *skip_macro_invocations* as it's more explicit, despite being longer.

ytmimi replied:
If you're leaning towards `skip_macro_invocations` then let's move forward with that one. I also wonder if we could update the help text here and in the Configurations.md file to be something like:

> rustfmt will skip formatting macro invocations for macros who's names appear in this list. set this value to "*" to prevent rustfmt from formatting all macros.
>
> Note: This option does not control formatting macro definitions.

Maybe just include the note in the Configurations.md file?

ytmimi reviewed (commented):
@tommilligan Thanks so much for getting this PR in! The approach you took looks great to me! Just pointed out a few minor things. Let me know if anything is unclear.

I'm still wondering what the best way to expose this new option to users is. As I mentioned in a previous comment, the `Ignore` list can only be set via the `rustfmt.toml` and I wonder if we should take a similar approach with this new option.

calebcartwright commented:
Thanks so much @tommilligan!

I wanted to echo @ytmimi's [question](https://github.com/rust-lang/rustfmt/pull/5347#discussion_r878895623) about definitions vs. bodies. The latter was what I understood the original scope to be, and which I believe to be the more important of two. I feel like definitions can be more easily handled via the outer skip attribute (since definition occurs only once), whereas calls can be scattered throughout the codebase.

Additionally, while pondering the above and the implications for the config name, I was also wondering if we should provide a mechanism to disable _all_ calls in a simple manner, e.g. `["*"]`?

tommilligan commented:
> Additionally, while pondering the above and the implications for the config name, I was also wondering if we should provide a mechanism to disable _all_ calls in a simple manner, e.g. `["*"]`?

@calebcartwright this is now implemented! Nice idea.

> definitions vs. bodies

ref: this thread, I think this may be a question of renaming the option to clarify, if you have any opinion: https://github.com/rust-lang/rustfmt/pull/5347#discussion_r894704462

tommilligan commented on src/skip.rs:
```diff
@@ -7,6 +7,7 @@ use rustc_ast_pretty::pprust;
 /// by other context. Query this context to know if you need skip a block.
 #[derive(Default, Clone)]
 pub(crate) struct SkipContext {
+    pub(crate) all_macros: bool,
     macros: Vec<String>,
```
I don't really like this structure, I think it should be refactored into something like

```rust
enum SkipMacroContext {
    All,
    Named(HashSet<MacroName>),
}

pub(crate) struct SkipContext {
    macros: SkipMacroContext,
    attributes: HashSet<String>,
}
```

but thought that was probably out of scope for this PR. Happy to open a followup afterwards, or include as an additional commit if you think it's okay to add here.

ytmimi replied:
I think the PR is in a pretty good state and I'm leaning towards making this change in a follow up PR. If you feel strongly about it though and don't think it will add a lot of extra work to the current PR, then I'll leave that call up to you

tommilligan commented:
@ytmimi thanks for your comments, they were super helpful. I've implemented or commented on all your points, so ready for another review.

ytmimi commented:
@tommilligan Thanks for taking the time to address all my comments. I've been away on a trip for the last two weeks so I haven't had much time for the project, but I'm back now and I plan on going through this again early next week (likely tomorrow!)

calebcartwright commented:
Will just add that I agree with `skip_macro_invocations` as the name (and have updated the PR title to that effect), agree with the proposed approach in https://github.com/rust-lang/rustfmt/pull/5347#discussion_r895000579 but doing so in a follow up, and also agree with the additional test cases outlined in https://github.com/rust-lang/rustfmt/pull/5347#pullrequestreview-1005832979

Everything else looks to be in good order to me :+1:

tommilligan commented:
Updated to:

- Expand the documentation
- Rename the option to `skip_macro_invocations`
- Add test cases as suggested

@ytmimi thanks for your test case list - there were some that I disagreed with:

> - calling a macro with a fully qualified path (e.g. self::items), where items is in the ignore list -> Expect that self::items would be ignored
This isn't the current behaviour - and I don't think it should be. As written, the feature is based purely on the literal invocation name, not on resolving the macro's identity to some more rigorous level - and I think that's fine. I think it's out of scope to handle path resolution.

We could do a really simple pass of 'remove all but the last path segment of the invocation name`, but then you couldn't do a more qualified match for e.g. `pretty_assertions::assert_eq`. You also wouldn't be able to selectively ignore just some macro invocations with the same name, but qualified by different paths.

> - aliasing a macro so that it matches up with the name in the ignore list (e.g. self::items as renamed_items), where renamed_items is in the ignore list. -> Expect the aliased macro to be ignored

This is implicitly tested in the different cases I added, because the `use` statement doesn't impact this rule, it's just the invocation name.

ytmimi commented:
> > * calling a macro with a fully qualified path (e.g. self::items), where items is in the ignore list -> Expect that self::items would be ignored
> >   This isn't the current behaviour - and I don't think it should be. As written, the feature is based purely on the literal invocation name, not on resolving the macro's identity to some more rigorous level - and I think that's fine. I think it's out of scope to handle path resolution.
> 
> We could do a really simple pass of 'remove all but the last path segment of the invocation name`, but then you couldn't do a more qualified match for e.g. `pretty_assertions::assert_eq`. You also wouldn't be able to selectively ignore just some macro invocations with the same name, but qualified by different paths.

That's a good point. I'm happy to go along with this behavior, and not ignore fully qualified macros if they aren't explicitly listed. I don't want to expand the scope of the current PR, and maybe this next idea needs to be discussed in more detail, but I wonder if we also need to consider the wild card case for fully qualified paths? Would `"self::*"` in the ignore list prevent `self::item` and `self::item as renamed` from being formatted? Again, I don't want to expand the scope of what we have right now, but I could see users asking for something like this in the future.

> 
> > * aliasing a macro so that it matches up with the name in the ignore list (e.g. self::items as renamed_items), where renamed_items is in the ignore list. -> Expect the aliased macro to be ignored
> 
> This is implicitly tested in the different cases I added, because the `use` statement doesn't impact this rule, it's just the invocation name.

Yup, I get that it's implicitly tested, but I think having at least one test where we're explicitly testing for it will make it clearer in the future that we've also considered this case.

tommilligan commented:
> ..., I wonder if we also need to consider the wild card case for fully qualified paths? Would `"self::*"` in the ignore list prevent `self::item` and `self::item as renamed` from being formatted? Again, I don't want to expand the scope of what we have right now, but I could see users asking for something like this in the future.

I think that's a reasonable feature request, and one that's compatible with the current design. As those patterns are invalid macro names, I think we're fine to give them special functionality in future without breaking existing behaviour.

I think that functionality would also be reasonable simple to add, but let's keep the scope down for now and wait for someone to ask!

> Yup, I get that it's implicitly tested, but I think having at least one test where we're explicitly testing for it will make it clearer in the future that we've also considered this case.

Fair enough - added a new test case file to demonstrate the current behaviour, and ensure use statements don't impact the behaviour (I guess basically a regression test actually)

ytmimi commented:
Thanks for adding the additional test! I'm looking to give this another review later this week.

tommilligan commented:
Rebased on master to resolve merge conflict.

calebcartwright commented on Configurations.md:
```diff
@@ -1014,6 +1014,60 @@ macro_rules! foo {
+
+#### `[]` (default):
+
+All macro invocations will be formatted.
```
I think it will probably also be worth mentioning and reinforcing rustfmt's macro formatting behavior anyway. For example, calls that use brace delimiters are essentially already skipped anyway, while calls with paren or bracket delims are attempted to be formatted if some other conditions hold true.

It may not make sense to try to articulate the totality of that behavior here, but if we can come up with a simple way to reference that we can (later) add a link that has more details

calebcartwright replied:
I actually went ahead and put up #5437 with a quick and dirty explanation that I think can suffice for now to give us something to point to here. Feel free to wordsmith, but I'd suggest we phrase this something like:

```suggestion
rustfmt will follow its standard approach to formatting macro invocations, and no macro invocations will be skipped based on their name. More information about rustfmt's standard macro invocation formatting behavior can be found in #5437
```

tommilligan replied:
Thanks for the suggestion and the writeup as well, that's nice and clear.

Updated the configurations doc

calebcartwright commented on Configurations.md:
```diff
@@ -1006,6 +1006,55 @@ macro_rules! foo {
 
+## `skip_macro_names`
+
+Skip formatting the bodies of macro invocations with the following names.
```
Perhaps this for the opening line?
```suggestion
Specify macro invocations that rustfmt should always skip, regardless of which delimiters are used.
```

ytmimi reviewed (approved):
Everything here looks good to me! I think we might consider the rewording proposed in https://github.com/rust-lang/rustfmt/pull/5347/files#r917441100, but I can't think of anything else to add here

calebcartwright reviewed (approved):
Going to go ahead and merge this, thanks for putting this together!

@ytmimi we can make any other changes, like tweaking the config text or adding some tests that use bracket delims in follow ups
